release/sil/sil.km.ggc/source/sil.km.ggc.model.ts

```diff
@@ -10,59 +10,71 @@
+        tokens.splice(i, 1);
+        i--;
+        continue;
+      }
 
-    // Certain punctuation marks should be considered a separate token from the word they're next to.
-    const punctuationMarks = ['«', '»', '$', '#' /* add extras here */];
-    const punctSplitIndices = [];
+      // Certain punctuation marks should be considered a separate token from the word they're next to.
+      const punctuationMarks = ['«', '»', '$', '#' /* add extras here */];
+      const punctSplitIndices = [];
 
-    // Find if and where each mark exists within the token
-    for(let i = 0; i < punctuationMarks.length; i++) {
-      const split = token.indexOf(punctuationMarks[i]);
-      if(split >= 0) {
-        punctSplitIndices.push(split);
+      // Find if and where each mark exists within the token
+      for(let i = 0; i < punctuationMarks.length; i++) {
+        const split = token.indexOf(punctuationMarks[i]);
+        if(split >= 0) {
+          punctSplitIndices.push(split);
+        }
       }
-    }
 
-    // Sort and pick the earliest mark's location.  If none exists, use -1.
-    punctSplitIndices.sort();
-    const splitPoint = punctSplitIndices[0] === undefined ? -1 : punctSplitIndices[0];
+      // Sort and pick the earliest mark's location.  If none exists, use -1.
+      punctSplitIndices.sort();
+      const splitPoint = punctSplitIndices[0] === undefined ? -1 : punctSplitIndices[0];
 
-    if(splitPoint > -1) {
-      const left = token.substring(0, splitPoint);  // (0, -1) => ''
-      const punct = token.substring(splitPoint, splitPoint+1);
-      const right = token.substring(splitPoint+1);  // Starting past the end of the string => ''
+      if(splitPoint > -1) {
+        const left = token.substring(0, splitPoint);  // (0, -1) => ''
+        const punct = token.substring(splitPoint, splitPoint+1);
+        const right = token.substring(splitPoint+1);  // Starting past the end of the string => ''
 
-      if(left) {
-        tokens.splice(i++, 0, left);
-      }
-      tokens.splice(i++, 1, punct);
-      if(right) {
-        tokens.splice(i, 0, right);
+        if(left) {
+          tokens.splice(i++, 0, left);
+        }
+        tokens.splice(i++, 1, punct);
+        if(right) {
+          tokens.splice(i, 0, right);
+        }
+        // Ensure that the next iteration puts `i` immediately after the punctuation token... even if
+        // there was a `right` portion, as it may have extra marks that also need to be spun off.
+        i--;
       }
-      // Ensure that the next iteration puts `i` immediately after the punctuation token... even if
-      // there was a `right` portion, as it may have extra marks that also need to be spun off.
-      i--;
     }
-   }
-   return tokens.map(function(token) {
+
+    let latestIndex = 0;
+    return tokens.map(function(token) {
+      const start = str.indexOf(token, latestIndex);
+      latestIndex = start + token.length;
```

Comment:
The key part of the fix:  we should track the furthest index reached while iterating over each token and skip past it when determining the original position of the next token to be processed.  This fixes handling for any duplicate-string cases.